File: 20-29/problem23.pl
package problem23;
use lib "../Utils";
require Math;

my $sum = find_sum();
print_result($sum);

sub find_sum {
  my $sum = 0;
  my %abundant_numbers = ();
  my $limit = 28123;
  for (my $ i = 12; $i <= $limit; $i++){
    if(Math::sum_of_proper_divisors($i) > $i){
      #Abundant number
      $abundant_numbers{$i} = $i;
    }
  }

  my %valid_numbers = ();
  for my $i (keys %abundant_numbers){
    for my $j (keys %abundant_numbers){
      if($i + $j <= limit){
        $valid_numbers{$i + $j} = 1;
      }else{
        last;
      }
    }
  }

  while(($number, $isvalid) = each %valid_numbers){
    if($isvalid){
      $sum += $number;
    }
  }

  return $sum;
}

sub print_result {
  my $res = shift;
  print "The sum of all the positive integers which cannot be written as the sum of two abundant numbers is $res.\n";
  return 1;
}

File: 20-29/problem23.cpp
#include <iostream>
#include <cmath>
#include <vector>
#include <map>
#include "../Utils/chrono.cpp"

bool is_abundant(int n);

int main(void){
	Chrono c;
	c.start();
	
	std::vector<int> abundant_numbers;
	const int n = 28123;
	for(int i = 2; i <= n; i++){
		if(is_abundant(i)){
			abundant_numbers.push_back(i);
		}
	}
	
	std::vector<bool> can_be_written_as_abundant(n);
	for(int i = 0; i < abundant_numbers.size(); i++){
		for(int j = i; j < abundant_numbers.size(); j++){
			if(abundant_numbers[i] + abundant_numbers[j] <= n){
				can_be_written_as_abundant[abundant_numbers[i] + abundant_numbers[j]] = true;
			}
		}
	}
	
	int sum = 0;
	for(int i = 0; i < n; i++){
		if(!can_be_written_as_abundant[i]){
			sum += i;
		}
	}

	c.stop();
	
	std::cout << "The sum of all the positive integers which cannot be written as the sum of two abundant numbers is " << sum << std::endl;
	std::cout << "Found the answer in " << c.time() << " seconds" << std::endl;
	return 0;
}

bool is_abundant(int n){
	int sum = 0;
	for(int i = 1; i <= n/2; i++){
		if(n%i == 0){
			sum += i;
		}
	}
	return sum > n ? true : false;
}



